include/arch/arc/arch.h

```diff
@@ -79,87 +80,70 @@ extern "C" {
 		1 << (31 - __builtin_clz(x) + 1) : \
 		1 << (31 - __builtin_clz(x)))
 
-#if defined(CONFIG_USERSPACE)
 
+#if defined(CONFIG_USERSPACE)
 #define Z_ARCH_THREAD_STACK_RESERVED \
 	(STACK_GUARD_SIZE + CONFIG_PRIVILEGED_STACK_SIZE)
+#else
+#define Z_ARCH_THREAD_STACK_RESERVED (STACK_GUARD_SIZE)
+#endif
 
-#if CONFIG_ARC_MPU_VER == 2
 
-#define Z_ARCH_THREAD_STACK_DEFINE(sym, size) \
-	struct _k_thread_stack_element __noinit \
-		__aligned(POW2_CEIL(STACK_SIZE_ALIGN(size))) \
-		sym[POW2_CEIL(STACK_SIZE_ALIGN(size)) + \
-		+  STACK_GUARD_SIZE + CONFIG_PRIVILEGED_STACK_SIZE]
+/* MPUv2 requires size must be power of 2 and >= 2048 */
+#define Z_ARCH_THREAD_STACK_SIZE(size) POW2_CEIL(STACK_SIZE_ALIGN(size))
 
+#if  defined(CONFIG_USERSPACE) && CONFIG_ARC_MPU_VER == 2
+#define Z_ARCH_THREAD_STACK_ALIGN(size)	Z_ARCH_THREAD_STACK_SIZE(size)
 #define Z_ARCH_THREAD_STACK_LEN(size) \
-	    (POW2_CEIL(STACK_SIZE_ALIGN(size)) + \
-	     MAX(POW2_CEIL(STACK_SIZE_ALIGN(size)), \
-		 POW2_CEIL(STACK_GUARD_SIZE + CONFIG_PRIVILEGED_STACK_SIZE)))
-
-#define Z_ARCH_THREAD_STACK_ARRAY_DEFINE(sym, nmemb, size) \
-	struct _k_thread_stack_element __noinit \
-		__aligned(POW2_CEIL(STACK_SIZE_ALIGN(size))) \
-		sym[nmemb][Z_ARCH_THREAD_STACK_LEN(size)]
+		(Z_ARCH_THREAD_STACK_SIZE(size) + Z_ARCH_THREAD_STACK_RESERVED)
+/*
+ * for stack arrary, it has more strict requirement on
+ * size and address alignment which is decided by MPUv2
+ */
+#define Z_ARCH_THREAD_STACK_ARRAY_LEN(size) \
+		(Z_ARCH_THREAD_STACK_SIZE(size) + \
+		MAX(Z_ARCH_THREAD_STACK_SIZE(size), \
+		POW2_CEIL(Z_ARCH_THREAD_STACK_RESERVED)))
+#elif defined(CONFIG_MPU_STACK_GUARD) && CONFIG_ARC_MPU_VER == 2
+#define Z_ARCH_THREAD_STACK_ALIGN(size)	(STACK_ALIGN)
+#define Z_ARCH_THREAD_STACK_LEN(size) \
+		(size + Z_ARCH_THREAD_STACK_RESERVED)
+#define Z_ARCH_THREAD_STACK_ARRAY_LEN(size) \
+		(Z_ARCH_THREAD_STACK_SIZE(size) + \
```

Comment:
Is the power-of-two alignement for stack size required here? Isn't stack size aligment (STACK_SIZE_ALIGN) enough to guarantee the guard address and size match MPU alignment requirements?

Reply:
done